src/test/kotlin/com/github/avrokotlin/avro4k/io/AvroSingleObjectInputStreamTest.kt

```diff
@@ -0,0 +1,52 @@
+package com.github.avrokotlin.avro4k.io
+
+import com.github.avrokotlin.avro4k.Avro
+import com.github.avrokotlin.avro4k.decoder.FooString
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.equals.shouldBeEqual
+import kotlinx.serialization.Serializable
+import org.apache.avro.generic.GenericRecord
+import org.apache.avro.message.SchemaStore
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+
+class AvroSingleObjectInputStreamTest : StringSpec({
+
+   "AvroSingleObjectInputStream should read otherFooString from fooString bytes" {
+      val foo = FooString(str = "Bar")
+      val writerSchema = Avro.default.schema(FooString.serializer())
+      val schemaStore = SchemaStore.Cache().apply {
+         addSchema(writerSchema)
+      }
+      val readerSchema = Avro.default.schema(OtherFooString.serializer())
+
+      val baos = ByteArrayOutputStream()
+
+      Avro.default.openOutputStream(FooString.serializer()) {
+         this.encodeFormat = AvroEncodeFormat.SingleObject
+         this.schema = writerSchema
+      }.to(baos).write(foo).close()
+
+      val decodedBytes = baos.toByteArray()
+
+      val other: OtherFooString = Avro.default.openInputStream(OtherFooString.serializer()) {
+         decodeFormat = AvroDecodeFormat.SingleObject(
+            readerSchema = readerSchema,
+            schemaStore = schemaStore
+         )
+      }.from(ByteArrayInputStream(decodedBytes)).nextOrThrow()
+
+      // value is equal
+      other.str.shouldBeEqual(foo.str)
```

Comment:
Sorry for the late review, I think it's better to test serialize+deserialize inside the same method. I think you can serialize with avro4k then deserialize with official avro library, and do the reverse (serialize with official avro library then deserialize with avro4k). That way you are sure that avro4k is compatible with the avro official library 😄

Reply:
Hm. I do not really get your point. Do you mean The InputStreamTest and the OutputStreamTest should become one test class and use java->kotlinx and kotlinx->java de/serializations?

Wouldn't then a use-case based test that does not directly refer to one of the concrete stream implementations the better option? So on top level "SingleObjectSerializationTest" with everything related?

Or did I completly miss your intention?

Reply:
Assuming that I got this right: I created a use case test: https://github.com/toolisticon/avro4k/blob/feature/141_single_object/src/test/kotlin/com/github/avrokotlin/avro4k/UseCaseSingleObjectEncodingTest.kt

This uses FooString (from the avro4k decoder-test) and FooStringJava (a java specificRecord I generated with the official maven plugin) to test all interoperability cases:

kotlin -> kotlin
kotlin -> java
java -> kotlin
java -> java

It is green ... let me know if this is what you had in mind.
If so, we can remove/refactor the additional tests for the In/Out streams.

It might be a good idea to also implement the other use cases (aka Decoding Formats) in this way.

Reply:
Not really, I was talking about those tests:
- encoding data from pure java avro encoder (with the schema coming from Avro.schemaFor since we are not testing this here) then try to decode it using your new method in avro4k (GenericDatumWriter by example)
- and the reverse, encode data from your new method in avro4k (still using the same schema from avro4k) then decode it using pure java avro decoder (GenericDatumReader by example)
- In both tests, you want to check that the finally decoded data is the same as the previouslyencoded data

Why ? It prove that we are fully compliant with the pure java avro library. In your current tests, you are encoding and decoding the data only using your avro4k methods

Reply:
see https://github.com/avro-kotlin/avro4k/pull/142#discussion_r1304252441